LOEUF.pm

```diff
@@ -0,0 +1,202 @@
+
+=cut
+
+package LOEUF;
+
+use strict;
+use warnings;
+
+use Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin;
+
+use base qw(Bio::EnsEMBL::Variation::Utils::BaseVepTabixPlugin);
+
+use Scalar::Util qw(looks_like_number);
+
+sub new {
+  my $class = shift;
+
+  my $self = $class->SUPER::new(@_);
+
+  $self->expand_left(0);
+  $self->expand_right(0);
+
+  my $param_hash = $self->params_to_hash();
+  
+  # Test tabix
+  die "ERROR: tabix does not seem to be in your path\n" unless `which tabix 2>&1` =~ /tabix$/;
+
+  # Get file
+  die "ERROR: LOEUF file not provided or not found!\n" unless defined($param_hash->{file}) && -e $param_hash->{file};
+  $self->add_file($param_hash->{file});
+
+  # Get headers from file
+  my $headers;
+  open HEAD,"tabix -H $param_hash->{file} 2>&1 | ";
+  while(<HEAD>) {
+    chomp;
+    $_ =~ s/^\#//;
+    $headers = [split];
+  }
+  close HEAD;
+
+  # Compare indexes of expected and observed columns
+  die "ERROR: Could not read headers from $param_hash->{file}\n" unless defined($headers) && scalar @{$headers};
+  my @obs_columns = @{$headers}[1,30,64] ;
+  my @exp_columns = ("transcript", "oe_lof_upper", "gene_id"); 
+  my @missing_columns;
+  foreach my $index (0 .. $#exp_columns){
+    if ($obs_columns[$index] ne  $exp_columns[$index]){
+      push @missing_columns,$exp_columns[$index];
+    }
+  }
+  my $missing_columns_str =  join(", ", @missing_columns) if scalar @missing_columns; 
+  die "ERROR: Missing columns: $missing_columns_str\n" if defined($missing_columns_str);
+
+  # Check match_by argument and store on self
+  if(defined($param_hash->{match_by})) {
+    my $match_by = $param_hash->{match_by};
+    $self->{match_by} = $match_by;
+  }
+
+  else{
+    die "ERROR: Argument 'match_by' is undefined\n" ;
+  }
+
+  # Check assembly
+  my $assembly = $self->{config}->{assembly};
+  if ($assembly ne "GRCh37") {
+    die "ERROR: Assembly is not GRCh37, LOEUF only works with GRCh37. \n";
+  }
+
+  return $self;
+}
+
+sub feature_types {
+  return ['Transcript'];
+}
+
+sub get_header_info {
+  return { LOEUF => 'Loss-of-function observed/expected upper bound fraction'};
+}
+
+sub run {
+  my ($self, $tva) = @_;
+
+  return {} if grep {$_->SO_term eq 'downstream_gene_variant' || $_->SO_term eq 'upstream_gene_variant'} @{$tva->get_all_OverlapConsequences};
+  
+  my $vf = $tva->variation_feature;
+  my $transcript = $tva->transcript;
+  my $end = $vf->{end};
+  my $start = $vf->{start};
+  ($start, $end) = ($end, $start) if $start > $end;
+  
+  if ($self->{match_by} eq 'transcript'){
+    my ($res) = grep {
+    $_->{transcript_id}  eq $transcript->stable_id;
+    } @{$self->get_data($vf->{chr}, $start, $end)};
+    return $res ? $res->{result} : {};
+  }
+
+  elsif ($self->{match_by} eq 'gene'){
+    my @data = grep {
+    $_->{gene_id} eq $transcript->{_gene_stable_id}
+    } @{$self->get_data($vf->{chr}, $start, $end)};
+    # Get max loeuf value for the gene
+    my $max_loeuf_score = 0;
```

Comment:
@likhitha-surapaneni - can you switch this to report the minimum score for a gene
Low LOEUF scores indicate strong selection against predicted loss-of-function (pLoF) variation in a given gene, so changes are more likely to be damaging.